statistics/databuffer.go

```diff
@@ -0,0 +1,412 @@
+)
+
+//functions to sort data
+func (a buffer) Len() int           { return len(a) }
+func (a buffer) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a buffer) Less(i, j int) bool { return a[i].value < a[j].value }
+
+//Inserts data in a sorted order
+func (b *dataBuffer) Insert(value float64, ts time.Time) {
+	if len(b.data) < cap(b.data) {
+		element := &data{value: value, ts: ts}
+		b.data = append(b.data, element)
+		b.dataByTimestamp = append(b.dataByTimestamp, element)
+	} else {
+		b.dataByTimestamp[b.index].ts, b.dataByTimestamp[b.index].value = ts, value
+		b.index = (b.index + 1) % cap(b.data)
+	}
+	sort.Sort(b.data)
+}
+
+//Set flags to true to be able to configure process functions
+func SetFlags(stats []string) (flag Flag, err error) {
+	for _, stat := range stats {
+		switch stat {
+		case count:
+			flag.Count = true
+		case mean:
+			flag.Mean = true
+		case median:
+			flag.Median = true
+		case standarddeviation:
+			flag.StandardDeviation = true
+		case variance:
+			flag.Variance = true
+		case ninetyfifthpercentile:
+			flag.NinetyFifthPercentile = true
+		case ninetyninthpercentile:
+			flag.NinetyNinthPercentile = true
+		case secondpercentile:
+			flag.SecondPercentile = true
+		case ninthpercentile:
+			flag.NinthPercentile = true
+		case twentyfifthpercentile:
+			flag.TwentyFifthPercentile = true
+		case seventyfifthpercentile:
+			flag.SeventyFifthPercentile = true
+		case ninetyfirstpercentile:
+			flag.NinetyFirstPercentile = true
+		case ninetyeighthpercentile:
+			flag.NinetyEighthPercentile = true
+		case minimum:
+			flag.Minimum = true
+		case maximum:
+			flag.Maximum = true
+		case rangeval:
+			flag.Range = true
+		case mode:
+			flag.Mode = true
+		case kurtosis:
+			flag.Kurtosis = true
+		case skewness:
+			flag.Skewness = true
+		case sum:
+			flag.Sum = true
+		case trimean:
+			flag.Trimean = true
+		case quartilerange:
+			flag.QuartileRange = true
+		case firstquartile:
+			flag.FirstQuartile = true
+		case thirdquartile:
+			flag.ThirdQuartile = true
+		default:
+			err = fmt.Errorf("Unknown statistic received %T:", stat)
+			return
+		}
+	}
+	return
+}
+
+// Creates a metric for each statistic
+func createMetric(data interface{}, tags map[string]string) plugin.Metric {
+	return plugin.Metric{
+		Timestamp: time.Now(),
+		Tags:      tags,
+		Data:      data,
+	}
+}
+
+// Get tags which are start time and stop time
+func (b *dataBuffer) GetTags() map[string]string {
+	new := (b.index + len(b.data) - 1) % len(b.data)
+	old := b.index
+	oldTs := b.dataByTimestamp[old].ts
+	newTs := b.dataByTimestamp[new].ts
+	return map[string]string{"startTime": oldTs.String(), "stopTime": newTs.String()}
+}
+
+// Checks if flag is true and calls the appropriate process functions
+func (d *dataBuffer) GetStats(stats []string, ns []string) ([]plugin.Metric, error) {
+	var results []plugin.Metric
+	var metric plugin.Metric
+	nsPrefix := []string{"intel", "statistics"}
```

Comment:
let's not use this prefix. Otherwise, the namespace look weird. e.g.:

/intel/statistics/intel/psutil/vm/total/ninetyfirstpercentile|0|
/intel/statistics/intel/psutil/load/load1/maximum|1.58